File: 0486-predict-the-winner/0486-predict-the-winner.cpp
class Solution {
public:
    bool predictTheWinner(vector<int>& nums) {
        vector<vector<int>>memo(nums.size() , vector<int>(nums.size() , -1));

        int res = solve(0,nums.size() -1 , nums , memo);
        return res >= 0 ? true : false; 

    }
private:
    int solve(int i , int j , vector<int> & nums, vector<vector<int>>& memo){
        if(i == j){
            return nums[i];
        }

        if(memo[i][j] != -1){
            return memo[i][j];
        }

        int left = nums[i] - solve(i+1 , j , nums,memo);
        int right = nums[j]-solve(i , j -1 , nums , memo);

        memo[i][j] = max(left , right);

        return memo[i][j];
    }
};

File: 0486-predict-the-winner/0486-predict-the-winner.py
class Solution:
    def predictTheWinner(self, nums: List[int]) -> bool:
        memo = {}

        def maxDiff(i , j):
            if i == j:
                return nums[i]
            
            if (i , j) in memo:
                return memo[(i,j)]
            

            left = nums[i] - maxDiff(i+1 , j)
            right = nums[j] - maxDiff(i , j-1)
        
            memo[(i,j)] = max(left , right)
        
            return memo[(i,j)]
        
        return maxDiff(0,len(nums) -1) >=0
        
